I will fix this 
```
level=info msg="Synchronized ENI information" numInstances=1 numSecurityGroups=2 numSubnets=1 numVPCs=1 subsys=eni
level=info msg="Synchronized ENI information" numInstances=1 numSecurityGroups=2 numSubnets=1 numVPCs=1 subsys=eni
level=info msg="Synchronized ENI information" numInstances=1 numSecurityGroups=2 numSubnets=1 numVPCs=1 subsys=eni
level=info msg="Releasing excess IPs from node" available=18 excess=5 excessIps="[10.10.10.7 10.10.111.163 10.10.61.225 10.10.82.173 10.10.53.94]" instanceID=i-testNodeManagerReleaseAddress-1 name=node3 releasing="[10.10.10.7 10.10.111.163 10.10.61.225 10.10.82.173 10.10.53.94]" selectedInterface=75e5dd75-05f7-462c-aba6-33d5d8b00487 selectedPoolID=s-1 subsys=ipam used=10
level=info msg="Synchronized ENI information for the corresponding instance" instance=i-testNodeManagerReleaseAddress-1 numSecurityGroups=2 numSubnets=1 numVPCs=1 subsys=eni
    node_manager_test.go:473: 
        ... obtained int = 13
        ... expected int = 15
        
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃   FAIL  package: github.com/cilium/cilium/pkg/ipam   ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
--- FAIL: Test (13.48s)
--- FAIL: Test/IPAMSuite (13.48s)
--- FAIL: Test/IPAMSuite/TestNodeManagerReleaseAddress (6.04s)
level=info msg="Discovered new CiliumNode custom resource" name=node3 subsys=ipam
level=warning msg="Unable to compute pending pods, will not surge-allocate" error="pod store uninitialized" instanceID= name=node3 subsys=ipam
level=info msg="Resolving IP deficit of node" available=0 availableForAllocation=2048 emptyInterfaceSlots=0 instanceID= maxIPsToAllocate=19 name=node3 neededIPs=15 remainingInterfaces=0 selectedInterface= selectedPoolID=global subsys=ipam used=0
level=warning msg="Unable to compute pending pods, will not surge-allocate" error="pod store uninitialized" instanceID= name=node3 subsys=ipam
level=info msg="Resolving IP deficit of node" available=19 availableForAllocation=2029 emptyInterfaceSlots=0 instanceID= maxIPsToAllocate=8 name=node3 neededIPs=4 remainingInterfaces=0 selectedInterface= selectedPoolID=global subsys=ipam used=19
level=info msg="Releasing excess IPs from node" available=27 excess=8 excessIps="[18 17 16 15 14 13 20 19]" instanceID= name=node3 releasing="[18 17 16 15 14 13 20 19]" selectedInterface= selectedPoolID=global subsys=ipam used=10
    node_manager_test.go:505: 
        ... obtained int = 19
        ... expected int = 18
```

--- a/Documentation/network/concepts/ipam/eni.rst
+++ b/Documentation/network/concepts/ipam/eni.rst
@@ -310,13 +310,26 @@ calculation is performed:
 
 .. code-block:: go
 
-     spec.ipam.pre-allocate - (len(spec.ipam.available) - len(status.ipam.used))
+     availableIPs := len(spec.ipam.pool)
+     neededIPs = max(spec.ipam.pre-allocate - (availableIPs - len(status.ipam.used)), spec.ipam.min-allocate - availableIPs)
+     if spec.ipam.max-allocate > 0 {
+      neededIPs = min(max(spec.ipam.max-allocate - availableIPs, 0), neededIPs)
+     }
 
 For excess IP calculation:
 
 .. code-block:: go
 
-     (len(spec.ipam.available) - len(status.ipam.used)) - (spec.ipam.pre-allocate + spec.ipam.max-above-watermark)
+     availableIPs := len(spec.ipam.pool)
+     upperBound := spec.ipam.min-allocate + spec.ipam.max-above-watermark
+     switch {
+     case availableIPs <= upperBound:
+       excessIPs = 0
+     case len(status.ipam.used) <= upperBound && len(status.ipam.used) + spec.ipam.pre-allocate <= upperBound:
+       excessIPs = availableIPs - upperBound
+     default:
+       excessIPs = max(availableIPs - len(status.ipam.used) - upperBound, 0)
+     }
 
 Upon detection of a deficit, the node is added to the list of nodes which
 require IP address allocation. When a deficit is detected using the interval

--- a/pkg/aws/eni/node_manager_test.go
+++ b/pkg/aws/eni/node_manager_test.go
@@ -437,7 +437,7 @@ func (e *ENISuite) TestNodeManagerReleaseAddress(c *check.C) {
 	c.Assert(node.Stats().UsedIPs, check.Equals, 10)
 
 	// Trigger resync manually, excess IPs should be released
-	// 10 used + 3 pre-allocate + 2 max-above-watermark => 15
+	// 10 used + 2 pre-allocate + 3 max-above-watermark => 15
 	node = mngr.Get("node3")
 	eniNode, castOK := node.Ops().(*Node)
 	c.Assert(castOK, check.Equals, true)
@@ -470,7 +470,7 @@ func (e *ENISuite) TestNodeManagerReleaseAddress(c *check.C) {
 	c.Assert(testutils.WaitUntil(func() bool { return reachedAddressesNeeded(mngr, "node3", 0) }, 5*time.Second), check.IsNil)
 	node = mngr.Get("node3")
 	c.Assert(node, check.Not(check.IsNil))
-	c.Assert(node.Stats().AvailableIPs, check.Equals, 15)
+	c.Assert(node.Stats().AvailableIPs, check.Equals, 13)
 	c.Assert(node.Stats().UsedIPs, check.Equals, 10)
 }
 

--- a/pkg/ipam/node.go
+++ b/pkg/ipam/node.go
@@ -335,6 +335,12 @@ func calculateExcessIPs(availableIPs, usedIPs, preAllocate, minAllocate, maxAbov
 		if availableIPs <= (minAllocate + maxAboveWatermark) {
 			return 0
 		}
+
+		// if usedIPs+preAllocate not over minAllocate + maxAboveWatermark, only care
+		// the ips out of minAllocate + maxAboveWatermark
+		if (usedIPs + preAllocate) <= (minAllocate + maxAboveWatermark) {
+			return availableIPs - minAllocate - maxAboveWatermark
+		}
 	}
 
 	// Once above the minimum allocation level, calculate based on

--- a/pkg/ipam/node_manager_test.go
+++ b/pkg/ipam/node_manager_test.go
@@ -502,7 +502,7 @@ func (e *IPAMSuite) TestNodeManagerReleaseAddress(c *check.C) {
 	c.Assert(testutils.WaitUntil(func() bool { return reachedAddressesNeeded(mngr, "node3", 0) }, 5*time.Second), check.IsNil)
 	node = mngr.Get("node3")
 	c.Assert(node, check.Not(check.IsNil))
-	c.Assert(node.Stats().AvailableIPs, check.Equals, 18)
+	c.Assert(node.Stats().AvailableIPs, check.Equals, 19)
 	c.Assert(node.Stats().UsedIPs, check.Equals, 10)
 }
 

--- a/pkg/ipam/node_test.go
+++ b/pkg/ipam/node_test.go
@@ -39,7 +39,7 @@ var neededDef = []testNeededDef{
 var excessDef = []testExcessDef{
 	{0, 0, 0, 16, 0, 0},
 	{15, 0, 8, 16, 8, 0},
-	{17, 0, 8, 16, 0, 9}, // 17 used, 8 pre-allocate, 16 min-allocate => 1 excess
+	{17, 0, 8, 16, 0, 1}, // 17 used, 8 pre-allocate, 16 min-allocate => 1 excess
 	{20, 0, 8, 16, 4, 0}, // 20 used, 8 pre-allocate, 16 min-allocate, 4 max-above-watermark => 0 excess
 	{21, 0, 8, 0, 4, 9},  // 21 used, 8 pre-allocate, 4 max-above-watermark => 9 excess
 	{20, 0, 8, 20, 8, 0},

